Added high-content benchmarks and fixed starvation bugs

spencerkimball opened the pull request:
The bank benchmark was never doing anything before now as there
was no money in any of the accounts, so nothing would transfer.

Added high-contention workloads to the benchmarks and in the
process discovered starvation scenarios and an outright bug.

The starvation scenario was encountered when a batch containing
writes and an end transaction failed due to a serialized txn
having `Timestamp` != `OrigTimestamp`. In that event, we discard
the entire batch engine, meaning the intents for the writes are
discarded with everything else.

This change commits the results of a batch which fails with
EndTransaction.

There was also a bug when restarting a transaction. The replay
of the `BeginTransaction` statement was incorrectly using the
value of the transaction record instead of the transaction
passed in with the batch request. In the event that the txn
was pushed or aborted during the restart, this would cause
various problems when commit time comes.

The 2, 4, and 8 account benchmarks are pretty slow, but
optimizing those is the next step.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5257)

<!-- Reviewable:end -->

spencerkimball committed:
Added high-content benchmarks and fixed starvation bugs

The bank benchmark was never doing anything before now as there
was no money in any of the accounts, so nothing would transfer.

Added high-contention workloads to the benchmarks and in the
process discovered starvation scenarios and an outright bug.

The starvation scenario was encountered when a batch containing
writes and an end transaction failed due to a serialized txn
having `Timestamp` != `OrigTimestamp`. In that event, we discard
the entire batch engine, meaning the intents for the writes are
discarded with everything else.

This change commits the results of a batch which fails with
EndTransaction.

There was also a bug when restarting a transaction. The replay
of the `BeginTransaction` statement was incorrectly using the
value of the transaction record instead of the transaction
passed in with the batch request. In the event that the txn
was pushed or aborted during the restart, this would cause
various problems when commit time comes.

The 2, 4, and 8 account benchmarks are pretty slow, but
optimizing those is the next step.

petermattis commented:
@tschottdorf should look at the `storage` changes. The change to the bank benchmark looks fine modulo the account selection.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/bank_test.go, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuLv5mpbWxHy2-9UQu:-KCuLv5nA-qWJM2PuqDO:-614971347) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L65)):*
I usually do this on the command line via the `-cpu` flag. I think specifying it like this prohibits using the `-cpu` flag to change the setting. 

---

*[sql/bank_test.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuM8owRzagY1ui6IS8:-KCuM8owRzagY1ui6IS9:1090376001) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L69)):*
Why did you change this? The code was previously selecting to random accounts, guaranteeing that they were different. Now it is possible for `to == from` after randomly selecting `to`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-:-KCuMNOE3yZiZFBOJuED:-1420348923)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/bank_test.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuZ9K2EYzlfPo9jL00:-KCuZ9K2EYzlfPo9jL01:2142027629) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L89)):*
can you make this an argument passed to the function?

---

*[storage/replica.go, line 1514 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuY7PYMQxXcyb8OxGx:-KCuY7PYMQxXcyb8OxGy:-2143481583) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica.go#L1514)):*
cc https://github.com/cockroachdb/cockroach/issues/3037

---

*[storage/replica_command.go, line 325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYDe0dvLTrLUY8JJA:-KCuYDe0dvLTrLUY8JJB:1981847239) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_command.go#L325)):*
this comment is inaccurate now

---

*[storage/replica_test.go, line 2859 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYbydZjhJAgaa4dEx:-KCuYbydZjhJAgaa4dEy:-503837129) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_test.go#L2859)):*
nit: print the detail here rather than `pErr`

---

*[storage/replica_test.go, line 2881 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYhK2tjTki3j_v96O:-KCuYhK2tjTki3j_v96P:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_test.go#L2881)):*
ditto

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/bank_test.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuoFa13yYRYqMqoFxW:-KCuoFa2ZGwZ_2D5V5gn:1898280895) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L28)):*
Time to nuke this flag as well? Seems like a wrinkle we don't really use - should pick the amount randomly based on the average available per account.

---

*[storage/replica.go, line 1510 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCurDyqfpOkVB9UUuBI:-KCurDyqfpOkVB9UUuBJ:-252567875) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica.go#L1510)):*
This needs a test.

---

*[storage/replica_command.go, line 325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYDe0dvLTrLUY8JJA:-KCup0Fcnk8Z58cDApZ6:-1677130893) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_command.go#L325)):*
Since this is what fixed the previous behavior, should also add a comment as to why we don't allow the write.
Actually, the right thing to do should be updating our txn from the persisted proto, and then writing the new entry.

---

*[storage/replica_test.go, line 2826 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuqZMKS2Lwsf_tzs3F:-KCuqZMKS2Lwsf_tzs3G:-1619585554) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_test.go#L2826)):*
s/\.$/, preventing regression of a bug in which which blindly overwrote the transaction record on BeginTransaction./

---

*[storage/replica_test.go, line 2859 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYbydZjhJAgaa4dEx:-KCupYNRm2j2ifxn1Sls:243669375) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_test.go#L2859)):*
The detail doesn't actually contain the whole message, you want `pErr`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/bank_test.go, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuLv5mpbWxHy2-9UQu:-KCur904EDWkz80nIH35:-1391749973) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L65)):*
If the parallelism is set to 1, for example, nothing makes sense in terms of the benchmark. You'd get zero contention. It's crucial we have a set level of parallelism here. Should I just assume it will be run with `-cpu 4`?

---

*[sql/bank_test.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuM8owRzagY1ui6IS8:-KCusQxDqGMp1HUJGWl7:-1823842311) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L69)):*
The previous code was biased towards selecting `*numAccounts - 1`, which is not correct if trying to select a uniformly random account. It's not possible to select `to == from` with this code. The `for` loop condition prevents it.

---

*[sql/bank_test.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuZ9K2EYzlfPo9jL00:-KCusxgfZh7uZbKDgGpx:5173665) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L89)):*
OK

---

*[storage/replica.go, line 1510 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCurDyqfpOkVB9UUuBI:-KCv-1bE-krcxYLGMpJl:-502253829) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica.go#L1510)):*
Added

---

*[storage/replica_command.go, line 325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYDe0dvLTrLUY8JJA:-KCuvhP8_Nfrv_m2x033:959752069) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_command.go#L325)):*
Fixed comment and updated request txn from extant txn record and am now rewriting it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/bank_test.go, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuLv5mpbWxHy2-9UQu:-KCv0xTUEGH-FIzMZEOP:2062361210) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L65)):*
I believe the default is already the number of CPUs.

---

*[storage/replica_test.go, line 1668 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCv1JXU2ZrkGfc584W5:-KCv1JXU2ZrkGfc584W6:255173502) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L1668)):*
looks like this happens on a _not_ write intent error?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/bank_test.go, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuLv5mpbWxHy2-9UQu:-KCv0mMpPc7HBvWt3Gap:-1950035793) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L65)):*
`-cpu` defaults to the `runtime.NumCPU` which seems fine. I think it is useful to be able to set `-cpu 1` on the command line to see how fast the benchmark runs without contention.

---

*[sql/bank_test.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuM8owRzagY1ui6IS8:-KCv2PFsb6dGwV4Gb381:-1907736589) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L69)):*
It is subtle, but I think the probabilities work out for the previous code without the for-loop. You'll be setting `to` to `numAccounts - 1` with a probability of `((numAccounts - 1) / numAccounts) * (1 / (numAccounts - 1)) = 1 / numAccounts`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuYDe0dvLTrLUY8JJA:-KCv_cwjRjyhlN6oT9SW:1502799018) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/storage/replica_command.go#L325)):*
We shouldn't change the args, and we should pass the new txn out of `BeginTxn`:

```go
h.Txn = h.Txn.Clone()
h.Txn.Update(&txn)
reply.Txn = h.Txn // rest of the batch should use the new txn
```

---

*[storage/replica_test.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCvaaBxB0zFUCTnINta:-KCvaaBxB0zFUCTnINtb:1557663714) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L1666)):*
How do you guarantee that this doesn't just push the intent?

---

*[storage/replica_test.go, line 2871 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCvag5B0638Pg6of8iP:-KCvag5B0638Pg6of8iQ:-1980130881) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L2871)):*
should also test what I suggested in the comment above: the retrying batch uses the updated txn returned by `BeginTransaction`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 325 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCxXat1Xh2Q5k8z6Wwy:-KCxXat1Xh2Q5k8z6Wwz:-424037333) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_command.go#L325)):*
s/, updated/ updated/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sql/bank_test.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCuM8owRzagY1ui6IS8:-KD-Ji6aaohYHYbDLE2c:808642596) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d266955c81bb1cda380b337807b77b241145502/sql/bank_test.go#L69)):*
Ah, I didn't pay attention to the `*numAccounts - 1` part there.

---

*[storage/replica_test.go, line 1666 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCvaaBxB0zFUCTnINta:-KD-Ms0SaZhV-q-m5epX:-1626306426) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L1666)):*
Good point, Changed sender to the range itself instead of using the store. The range won't try any sort of push txn or resolve intent.

---

*[storage/replica_test.go, line 1668 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCv1JXU2ZrkGfc584W5:-KD-N0bsZCUcVigRXkPj:-926045954) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L1668)):*
Fixed error message

---

*[storage/replica_test.go, line 2871 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KCvag5B0638Pg6of8iP:-KD-Q-joqY2sl_I-qP5s:-905058265) ([raw file](https://github.com/cockroachdb/cockroach/blob/72bc5783737f947720ef7cba379a562ae497da6c/storage/replica_test.go#L2871)):*
OK, think I understood what you're asking for and added a clause to the test.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[storage/replica_test.go, line 2929 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KD-WlyWEjLA6cArYMjI:-KD-WlyWEjLA6cArYMjJ:-1367828889) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0a16ebfcd592967de97195ff7c6d8c248af62/storage/replica_test.go#L2929)):*
https://xkcd.com/859/

---

*[storage/replica_test.go, line 2930 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KD-WrqttTniSXd7BTMY:-KD-WrqutOgw6ViIw5Dg:-560846978) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0a16ebfcd592967de97195ff7c6d8c248af62/storage/replica_test.go#L2930)):*
a missing `!`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-:-KD-X-iK8Rha_-Hk1xeS:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 2930 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257#-KD-WrqttTniSXd7BTMY:-KD1lXhhO-A9g4MOvU5M:-497468049) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d0a16ebfcd592967de97195ff7c6d8c248af62/storage/replica_test.go#L2930)):*
Fixed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5257)*
<!-- Sent from Reviewable.io -->
